### PR TITLE
Implement NonEmptyList#Collect

### PR DESCRIPTION
**Pull request by xavier-fernandez**
- "collect" method is really useful when wanting to do a transformation in some of the elements from a collection, applying a partial function. This PR implements this method both for `NonEmptyVector` and `NonEmptyList`.

### CODE REVIEW

**Comment by johnynek** [comment]
I'm weakly -1 on adding methods which are identical to doing `.toList.foo` or `toVector.foo`.

Methods that preserve nonemptiness are more interesting to me.

**Comment by xavier-fernandez** [comment]
What about returning: `Option[NonEmptyList[B]]`?

**Comment by kailuowang** [comment]
@xavier-fernandez, unless I am missing something, a `Option[NonEmptyList[B]]` is as powerful as a `List[B]` but less convenient. So you do want to return a `List[B]`. And if you going to lose the nonemptiness anyway why not `x.toList.collect`? To me it makes the lost of nonemptiness more explicit. Thus I favor that than `x.collect` that returns a `List`. 
The `filter` method we have there is an existing example of such methods that don't preserve nonemptines. @ceedubs do you happen to remember the incentive of adding that filter method?

**Comment by xavier-fernandez** [comment]
I should close the PR?

**Comment by ceedubs** [comment]
The main arguments that I can think of for having the `filter` method is that for `NonEmptyList` it can be implemented every so slightly more efficiently than `toList.filter` and that it may be slightly more convenient for users. I do recall at least one time that a coworker has asked me how they should filter a scalaz `NonEmptyList`, as it isn't necessarily obvious that you should go through `List`. I'm pretty neutral as to whether or not we include the `List`-like methods that don't offer anything specific to being non-empty.

**Comment by xavier-fernandez** [comment]
`NonEmptyList.collect` will have a better performance than `toList.collect`...

On the other hand,`NonEmptyVector.collect` will not yield any performance improvement.

**Comment by johnynek** [comment]
Okay. On this argument I'm +1 on NonEmptyList.collect but not
NonEmptyVector.

**Comment by xavier-fernandez** [comment]
If `NonEmptyList` is having a collect method, it would not be better that `NonEmptyVector` too, for consistency?

**Comment by johnynek** [comment]
I don't think so. Otherwise we'll have some other class justifying why we
should have yet another method.

Frankly, if consistency is the main driver I suggest we go back to only
adding methods if they preserve non-emptiness.

**Comment by xavier-fernandez** [comment]
Modified the PR so the `collect` method is only implemented on `NonEmptyList`

File: core/src/main/scala/cats/data/NonEmptyList.scala [hunk 1]
```diff
@@ -67,6 +67,30 @@ final case class NonEmptyList[+A](head: A, tail: List[A]) {
   }
 
   /**
+    * Builds a new collection by applying a partial function to
```
Comment (peterneyens) [inline core/src/main/scala/cats/data/NonEmptyList.scala]: Should we say "Builds a new `List` ..." ?

File: core/src/main/scala/cats/data/NonEmptyList.scala [hunk 2]
```diff
@@ -67,6 +67,30 @@ final case class NonEmptyList[+A](head: A, tail: List[A]) {
...
+    *
+    * {{{
+    * scala> import cats.data.NonEmptyList
+    * scala> val nel = NonEmptyList.of(1, 2, 3, 4, 5)
+    * scala> nel.collect { case v if v < 3 => v }
+    * res0: scala.collection.immutable.List[Int] = List(1, 2)
+    * scala> nel.collect {
+    *   case v if v % 2 == 0 => "even"
+    *   case _ => "odd"
+    *   }
```
Comment (peterneyens) [inline core/src/main/scala/cats/data/NonEmptyList.scala]: I think this should be : 

```scala
/**
 * ...
 * scala> nel.collect {
 *      |   case v if v % 2 == 0 => "even"
 *      |   case _ => "odd"
 *      | }
 * ...
 */
```

for the doctest to be generated correctly.

**Comment by xavier-fernandez** [comment]
@peterneyens amend done :)

**Comment by johnynek** [comment]
👍

**Comment by peterneyens** [comment]
You need to use exactly 
```
     | 
```
after a `scala> ` for the test to be picked up.

Right now the generated doctest is 

```scala
    property("example at line 203: nel.collect { case v if v < 3 => v }") = org.scalacheck.Prop.secure {
      sbtDoctestTypeEquals(nel.collect { case v if v < 3 => v })((nel.collect { case v if v < 3 => v }): scala.collection.immutable.List[Int])
      val actual = sbtDoctestReplString(nel.collect { case v if v < 3 => v })
      val expected = "List(1, 2)"
      (actual == expected) :| s"'$actual' is not equal to '$expected'"
    }
```

If you change it to 

```scala
   /**
    * scala> nel.collect {
    *      |   case v if v % 2 == 0 => "even"
    *      |   case _ => "odd"
    *      | }
    */
```

It will include this one as well : 

```scala
    property("example at line 205: nel.collect { ...") = org.scalacheck.Prop.secure {
      sbtDoctestTypeEquals(nel.collect {
  case v if v % 2 == 0 => "even"
  case _ => "odd"
})((nel.collect {
  case v if v % 2 == 0 => "even"
  case _ => "odd"
}): scala.collection.immutable.List[Int])
      val actual = sbtDoctestReplString(nel.collect {
  case v if v % 2 == 0 => "even"
  case _ => "odd"
})
      val expected = "List(odd, even, odd, even, odd)"
      (actual == expected) :| s"'$actual' is not equal to '$expected'"
    }
```

It would be handy if `sbt-doctest` gave some kind of error message if it is ignoring lines.

**Comment by xavier-fernandez** [comment]
@peterneyens will fix tomorrow morning, thank you :)

**Comment by xavier-fernandez** [comment]
Did an ammend fixing the DocTest generation :)

**Comment by kailuowang** [comment]
👍  merging with 2 sign-offs.

**Pull request merged**